pkg/gameserverallocations/controller.go

```diff
@@ -65,25 +71,100 @@ type Controller struct {
 	gameServerAllocationSynced cache.InformerSynced
 	gameServerAllocationGetter getterv1alpha1.GameServerAllocationsGetter
 	stop                       <-chan struct{}
-	allocationMutex            *sync.Mutex
 	workerqueue                *workerqueue.WorkerQueue
+	gsWorkerqueue              *workerqueue.WorkerQueue
 	recorder                   record.EventRecorder
+	readyGameServers           gameServerCacheEntry
+	topNGameServerCnt          int
+}
+
+// gameserver cache to keep the Ready state gameserver.
+type gameServerCacheEntry struct {
+	mu    sync.RWMutex
+	cache map[string]*v1alpha1.GameServer
+}
+
+// LoadOrStore returns data if exists or stores if it is not. Return the value and true if the data already exists.
+func (e *gameServerCacheEntry) LoadOrStore(key string, gs *v1alpha1.GameServer) (*v1alpha1.GameServer, bool) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.cache == nil {
+		e.cache = map[string]*v1alpha1.GameServer{}
+	} else if val, ok := e.cache[key]; ok {
+		return val, ok
+	}
+
+	e.cache[key] = gs.DeepCopy()
+	return e.cache[key], false
+}
+
+// Store saves the data in the cache.
+func (e *gameServerCacheEntry) Store(key string, gs *v1alpha1.GameServer) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.cache == nil {
+		e.cache = map[string]*v1alpha1.GameServer{}
+	}
+	e.cache[key] = gs.DeepCopy()
+}
+
+// Delete deletes the data. If it exists returns true.
+func (e *gameServerCacheEntry) Delete(key string) bool {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	ret := false
+	if e.cache != nil {
+		if _, ok := e.cache[key]; ok {
+			delete(e.cache, key)
+			ret = true
+		}
+	}
+
+	return ret
+}
+
+// Load returns the data from cache. It return true if the value exists in the cache
+func (e *gameServerCacheEntry) Load(key string) (*v1alpha1.GameServer, bool) {
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+	val, ok := e.cache[key]
+
+	return val, ok
+}
+
+// Range extracts data from the cache based on provided function f.
+func (e *gameServerCacheEntry) Range(f func(key string, gs *v1alpha1.GameServer) bool) {
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+	for k, v := range e.cache {
+		if !f(k, v) {
+			break
+		}
+	}
 }
 
 // findComparator is a comparator function specifically for the
 // findReadyGameServerForAllocation method for determining
 // scheduling strategy
 type findComparator func(bestCount, currentCount NodeCount) bool
 
+var allocationRetry = wait.Backoff{
+	Steps:    5,
```

Comment:
I am not sure making this configurable will be helpful without some guidance. Eventually it can be. 
Once everything is stable, we can try different values and find a good number.
I get this from [here](https://github.com/GoogleCloudPlatform/agones/blob/master/vendor/k8s.io/client-go/util/retry/util.go). I was trying different numbers. I agree that it should not be too high. I can set it to 3 for now. 
What do you think?

Reply:
That sounds fair - more than anything else, I just wanted to ask the question. I'm happy with 5 - seems like a legitimate number. I feel like I'd rather aim higher, and then maybe pull it down if need be,

@Kuqd curious - at some point, should we add a metric for this somewhere? How many retries?